cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1103 @@
+        RootAccessTrie { trie: primary_map }
+    }
+}
+
+impl RootAccessTrie {
+    /// Create an empty [`RootAccessTrie`] that requests nothing.
+    pub fn new() -> Self {
+        Self {
+            trie: Default::default(),
+        }
+    }
+}
+
+impl<T: Clone> RootAccessTrie<T> {
+    /// Union two [`RootAccessTrie`]s together.
+    /// The new trie requests the data from both of the original.
+    fn union(&self, other: &Self) -> Self {
+        let mut res = self.clone();
+        for (key, value) in &other.trie {
+            if let Some(existing) = res.trie.get(key) {
+                res.trie.insert(key.clone(), (*existing).union(value));
+            } else {
+                res.trie.insert(key.clone(), value.clone());
+            }
+        }
+        res
+    }
+}
+
+impl Default for RootAccessTrie {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: Clone> AccessTrie<T> {
+    /// Union two [`AccessTrie`]s together.
+    /// The new trie requests the data from both of the original.
+    fn union(&self, other: &Self) -> Self {
+        Self {
+            children: union_fields(&self.children, &other.children),
+            parents_required: self.parents_required || other.parents_required,
+            data: self.data.clone(),
+        }
+    }
+}
+
+impl AccessTrie {
+    /// A new trie that requests no data.
+    fn new() -> Self {
+        Self {
+            children: Default::default(),
+            parents_required: false,
+            data: (),
+        }
+    }
+}
+
+/// Computes an [`EntityManifest`] from the schema and policies.
+/// The policies must validate against the schema in strict mode,
+/// otherwise an error is returned.
+pub fn compute_entity_manifest(
+    schema: &ValidatorSchema,
+    policies: &PolicySet,
+) -> Result<EntityManifest, EntityManifestError> {
+    let mut manifest: HashMap<RequestType, RootAccessTrie> = HashMap::new();
+
+    // now, for each policy we add the data it requires to the manifest
+    for policy in policies.policies() {
+        // typecheck the policy and get all the request environments
+        let typechecker = Typechecker::new(schema, ValidationMode::Strict, policy.id().clone());
+        let request_envs = typechecker.typecheck_by_request_env(policy.template());
+        for (request_env, policy_check) in request_envs {
+            // match on the typechecking answer
+            let new_primary_slice = match policy_check {
+                PolicyCheck::Success(typechecked_expr) => {
+                    // compute the trie from the typechecked expr
+                    // using static analysis
+                    compute_root_trie(&typechecked_expr, policy.id())
+                }
+                PolicyCheck::Irrelevant(_) => {
+                    // this policy is ireelevant, so we need no data
+                    Ok(RootAccessTrie::new())
+                }
+
+                // TODO is returning the first error correct?
```

Comment:
Need advice on this TODO

Reply:
Returning the first error is fine, but returning all the errors would be more consistent with the rest of the validation APIs.

Re your second question: yes, validation does some additional steps (e.g., handling for templates). If you look at `validate` and `validate_policies` [here](https://github.com/cedar-policy/cedar/blob/672074429b67cf6ec2727c423409b4ac5ff73582/cedar-policy-validator/src/lib.rs#L110) you'll see that there are several pre-processing steps before we get to `typecheck_policy` (which is what calls `typecheck_by_request_env`). I think it'd be ideal to add you changes to the main validation path (if possible).

Reply:
I'm not sure how to add it to the validation path without it getting ugly- it would have to return both the current result and an optional entity manifest.
I'll keep this as-is for now.

Reply:
Ended up running full validation first before this code path.